[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148537
* __->__ #148377

### Summary
This PR adds `_scaled_dot_product_cudnn_attention` to DTensor ops and tests it with unit test. This should allow Context Parallel and Tensor Parallel to use cudnn SDPA.

### Test
`pytest test/distributed/tensor/test_attention.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

XilunWu committed:
[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

drisspg commented on test/distributed/tensor/test_attention.py:
```diff
@@ -86,6 +89,10 @@ def _test_ring_attention_sdpa(
         test_forward_only: bool,
     ) -> None:
+        # TODO: SDPBackend.CUDNN_ATTENTION does not support backward so far
+        if not test_forward_only and backend == SDPBackend.CUDNN_ATTENTION:
```
Just to confirm this is specific to the Dtensor impl not the op itself - since that does support bwd

XilunWu replied:
yes, the next step is to add backward DTensor impl.

drisspg commented on test/distributed/tensor/test_attention.py:
```diff
@@ -41,6 +42,8 @@
     backends.append(SDPBackend.FLASH_ATTENTION)
 if PLATFORM_SUPPORTS_MEM_EFF_ATTENTION:
     backends.append(SDPBackend.EFFICIENT_ATTENTION)
+if PLATFORM_SUPPORTS_CUDNN_ATTENTION:
```
We will be recommending only using this option on h100 and b200 btw

XilunWu replied:
Will `PLATFORM_SUPPORTS_CUDNN_ATTENTION` reflect this? if so, we may need to manually copy the existing `PLATFORM_SUPPORTS_CUDNN_ATTENTION` check into the test since we do want to run this unit test on CI instance.

XilunWu committed:
Update on "[dtensor] add aten._scaled_dot_product_cudnn_attention.default op support"


### Summary
This PR adds `_scaled_dot_product_cudnn_attention` to DTensor ops and tests it with unit test. This should allow Context Parallel and Tensor Parallel to use cudnn SDPA.

### Test
`pytest test/distributed/tensor/test_attention.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "broke lint [GH job link](https://github.com/pytorch/pytorch/actions/runs/13683650448/job/38261818684) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/9eef457c0241f87097a2ca7625f9961e31f3adcd) probably landrace" -c landrace
